mmf/trainers/core/evaluation_loop.py

```diff
@@ -17,33 +17,53 @@
 
 class TrainerEvaluationLoopMixin(ABC):
     def evaluation_loop(
-        self, loader, use_tqdm: bool = False, single_batch: bool = False
+        self, dataset_type: str, use_tqdm: bool = False, single_batch: bool = False
     ) -> Tuple[Dict[str, Any], Type[Meter]]:
         meter = Meter()
+        reporter = self.dataset_loader.get_test_reporter(dataset_type)
 
         with torch.no_grad():
             self.model.eval()
             disable_tqdm = not use_tqdm or not is_master()
-            combined_report = None
 
-            for batch in tqdm.tqdm(loader, disable=disable_tqdm):
-                report = self._forward(batch)
-                self.update_meter(report, meter)
+            while reporter.next_dataset(flush_report=False):
+                dataloader = reporter.get_dataloader()
+
+                combined_report = None
+                for batch in tqdm.tqdm(dataloader, disable=disable_tqdm):
+                    prepared_batch = reporter.prepare_batch(batch)
+                    model_output = self.model(prepared_batch)
+                    report = Report(prepared_batch, model_output)
 
-                # accumulate necessary params for metric calculation
-                if combined_report is None:
-                    combined_report = report
-                else:
-                    combined_report.accumulate_tensor_fields(
-                        report, self.metrics.required_params
-                    )
-                    combined_report.batch_size += report.batch_size
+                    self.update_meter(report, meter)
 
-                if single_batch is True:
-                    break
+                    # accumulate necessary params for metric calculation
+                    if combined_report is None:
+                        # make a copy of report since `reporter.add_to_report` will
+                        # change some of the report keys later
+                        combined_report = Report(report)
+                    else:
+                        combined_report.accumulate_tensor_fields(
+                            report, self.metrics.required_params
+                        )
+                        combined_report.batch_size += report.batch_size
 
-            combined_report.metrics = self.metrics(combined_report, combined_report)
-            self.update_meter(combined_report, meter, eval_mode=True)
+                    # Each node generates a separate copy of predict JSON from the report,
+                    # which will be used to evaluate dataset-level metrics
+                    # (such as mAP in object detection or CIDEr in image captioning)
+                    # Since `reporter.add_to_report` changes report keys (e.g. scores),
+                    # do this after `combined_report.accumulate_tensor_fields`
+                    reporter.add_to_report(report, self.model, master_only=False)
```

Comment:
Isn't this unnecessary calculation for dataset which don't actually output this as well as for metrics which don't need this?

Reply:
Yes, that's true. What do you feel is a better way? Shall I add "test_report" was a `required_params` for each metric, and only gather report if "test_report" is in `required_params`?

Reply:
(Added `__prediction_report__` in a metric's `required_params`)